### ISSUE TITLE
Does not search Chinese, Japanese characters

### ISSUE THREAD

**Issue by Aspire1Inspire2** [signalapp/Signal-Android#8629]
- [x] I have searched open and closed issues for duplicates
- [x] I am submitting a bug report for existing functionality that does not work as intended
- [x] I have read https://github.com/signalapp/Signal-Android/wiki/Submitting-useful-bug-reports
- [x] This isn't a feature request or a discussion topic

----------------------------------------

### Bug description
The new signal search feature.does not search Asian Characters. It only search sentence.

### Steps to reproduce
- Have some complete sentences made of Asian Characters.
- search for individual character
- search returns no result.

**Actual result:** 
- search for the whole sentence
- search return matching result.
- search return no result for characters or combination of characters

**Expected result:** 
Search should return result for any combination of asian characters.

### Device info
I beleieve this issue is general for all signal apps because the search is built to search only roman characters between spaces, etc.

### Link to debug log
N/A
Inherent search implementation limitation.

**Comment by Aspire1Inspire2** [comment]
With the latest update, it still doesn't search asian characters, only the first characters combination in a sentence. It still recognize whatever between commas or periods as a single word. Asian characters are not formed this way.

**Comment by Aspire1Inspire2** [comment]
Issue still persist.

**Comment by Aspire1Inspire2** [comment]
Issue still persist.

**Comment by Aspire1Inspire2** [comment]
Issue still persist to this date.

**Comment by MewX** [comment]
And, English inside Chinese sentences can't be searched.

E.g. 啦啦啦abc啦啦啦
You won't find "abc" in search history.

**Comment by o3661606** [comment]
Does it related to the FTS (full-text search) engine of sqlite?

**Comment by Aspire1Inspire2** [comment]
> Does it related to the FTS (full-text search) engine of sqlite?

Not sure what signal uses as backend database or how the search works, I have never read any source code.

But it seems to me that signal only search word defined/delimited by spaces. Asian language does not delimit word by space, but by human Brian.

**Comment by nathancchu** [comment]
Any updates on this? This is the only reason that keeps me using apps from Facebook and Line and WeChat

**Comment by MewX** [comment]
Yes, it is still relevant. I have the same feeling that the blocker is current token parsing logic for non-space-split languages.

**Comment by nathancchu** [comment]
It's still relevant and not fixed yet

**Comment by MewX** [comment]
Still need a fix here.

**Comment by cschanaj** [comment]
My guess is that SQLCipher has [removed ICU support](https://discuss.zetetic.net/t/sqlcipher-for-android-release-android-n-support/1465) back in 2016 and the built-in SQLite tokenize does not support CJK. Besides, it seems that [fts5 does not support CJK](https://stackoverflow.com/questions/52422437/) without building with ICU. 

Dropping my research here in case someone is looking. A possible roadmap to fixing this could be:

- [ ] Confirm SQLCipher `tokenize=icu` support, or use a custom build with ICU support; other options are
      - use `fts5 tokenize=icu` or `fts4 tokenize=unicode61`
      - use the workaround [described in this post](https://discuss.zetetic.net/t/workaround-for-using-androids-localized-collator-after-3-5-0-update/1515/2?u=developernotes)
- [ ] Tokenize search query with `BreakIteratorCompat` in [`SearchDatabase.java`](https://github.com/signalapp/Signal-Android/blob/main/app/src/main/java/org/thoughtcrime/securesms/database/SearchDatabase.java)
- [ ] Create new search database with fts4 tokenize=icu (if fts5 does not support CJK)
- [ ] Migrate existing virtual tables in [`SignalDatabaseMigrations.kt`](https://github.com/signalapp/Signal-Android/blob/main/app/src/main/java/org/thoughtcrime/securesms/database/helpers/SignalDatabaseMigrations.kt)

**Comment by cschanaj** [comment]
@signalapp full text search is not working as what users might expected for CJK (non-ascii) characters. This significantly impact non-westerners initiatives to move from alternatives which are potentially less secure/ violating user privacy. Is there some resource constraint or technical reasons this issue is marked as `wont fix`, seriously?

**Comment by logico-philosophical** [comment]
A small example should help:

* (English) All human beings are born free and equal in dignity and **rights**.
* (Simplified Chinese) 人人生而自由，在尊严和**权利**上一律平等。
* (Japanese) すべての人間は、生れながらにして自由であり、かつ、尊厳と**権利**とについて平等である。
* (Korean) 모든 인간은 태어날 때부터 자유로우며 그 존엄과 **권리**에 있어 동등하다.

Search for rights/权利/権利/권리 in each sentences. The search doesn't work for simplified Chinese and Japanese versions. Although the search does work for the Korean version, this can be a problem for Korean too because Koreans may omit a lot of spaces when sending informal messages.

It seems that the search fails because the full-text search functionality of the SQLite FTS5 table can't search for something in the middle of a text not prefixed by delimiters like spaces. For example, if you search instead for `尊厳` in the Japanese text above the search does work, beause `尊厳` is prefixed by the delimiter `、`. Since Chinese and Japanese languages rarely use spaces, most of the time the search doesn't work for these languages.

The following pieces of code seem to be responsible for this behavior:

https://github.com/signalapp/Signal-Android/blob/3869de414fa882bf4b53473f45e22cf6b3a2e3d8/app/src/main/java/org/thoughtcrime/securesms/database/SearchTable.kt#L55-L76

Note that SQLite's `MATCH` operator is being used to perform the search.

https://github.com/signalapp/Signal-Android/blob/3869de414fa882bf4b53473f45e22cf6b3a2e3d8/app/src/main/java/org/thoughtcrime/securesms/database/SearchTable.kt#L151-L165

The `createFullTextSearchQuery` function transforms search queries like `尊厳 権利` into `"尊厳"* "権利"*`, which in turn produces SQL queries like `SELECT ... WHERE $MMS_FTS_TABLE_NAME MATCH '"尊厳"* "権利"*' AND ...`.

**Comment by keikhcheung** [comment]
The problem still exists. I have noticed that if the word is at the start of a sentence, right after a space, or right after a punctuation mark (full-width or half-width), the word can be found.

If the word is in the middle of a sentence without preceding spaces or punctuation marks, the word won't be returned as a matching result.

This confirms what @logico-philosophical wrote above on 9 Jan 2023.

This same issue also exists on iOS. On Signal Desktop for Mac, it is working without any problem, though.
